src/devices/machine/qvt102.cpp

```diff
@@ -0,0 +1,692 @@
+	uint8_t attr = 0;
+
+	for (int x = 0; x < x_count; x++)
+	{
+		uint16_t mem = (ma + x) & 0x7ff;
+		uint8_t chr = m_vram[mem];
+
+		// address lines 10 and 11 for the character rom get a special handling
+		int a10 = (chr & 0x40) || ((chr & 0xe0) == 0);
+		int a11 = 0;
+
+		switch (m_jumper->read() & 0x30)
+		{
+		case 0x00: a11 = 0; break;    // US
+		case 0x10: a11 = a10; break;  // German
+		case 0x20: a11 = !a10; break; // UK
+		}
+
+		uint16_t gfx = m_char_rom[(a11 << 11) | (a10 << 10) | ((chr << 4) & 0x3f0) | ra] << 1;
+
+		// d0 for drawing codes (avoids gap)
+		if (chr >= 0x80 && chr <= 0x9f)
+			gfx |= BIT(gfx, 1);
+
+		// half intensity is defined per character
+		int half = BIT(chr, 7);
+
+		// check for new attribute
+		if (chr >= 0x90 && chr <= 0x9f)
+		{
+			attr = chr;
+			half ^= BIT(m_jumper->read(), 6);
+		}
+
+		// draw 9 pixels of the character
+		for (int i = 0; i < 9; i++)
+		{
+			// from schematics
+			bool p = !(BIT(m_latch, 3) & BIT(attr, 1)); // blink
+			p = !(p & BIT(gfx, i) & (!BIT(attr, 0))); // rom/blank
+			p = !(p & !((BIT(attr, 3) & (ra == 11)))); // underline
+			p = p ^ (BIT(attr, 2) ^ ((x == cursor_x) | BIT(m_latch, 2))); // reverse
+
+			bitmap.pix32(y, x*9 + (8-i)) = palette[p ? 2 - half : 0];
+		}
+
+	}
+}
+
+static const gfx_layout char_layout =
+{
+	8,12,
+	RGN_FRAC(1,1),
+	1,
+	{ 0 },
+	{ 0, 1, 2, 3, 4, 5, 6, 7 },
+	{ 0*8, 1*8, 2*8, 3*8, 4*8, 5*8, 6*8, 7*8, 8*8, 9*8, 10*8, 11*8 },
+	8*16
+};
+
+static GFXDECODE_START(chars)
+	GFXDECODE_ENTRY("chargen", 0, char_layout, 0, 1)
+GFXDECODE_END
+
+
+//**************************************************************************
+//  MACHINE EMULATION
+//**************************************************************************
+
+WRITE8_MEMBER(qvt102_device::latch_w)
+{
+	// 7------- host to keyboard
+	// -6------ aux print
+	// --5----- ctc cs0
+	// ---4---- dtri
+	// ----3--- blink
+	// -----2-- rv
+	// ------1- eia/cl
+	// -------0 aux en
+
+	m_latch = data;
+
+	// jumper w3 enables this connection
+	if (machine().phase() != machine_phase::INIT)
+	{
+		if (BIT(m_jumper->read(), 2))
+		{
+			fprintf(stderr, "######## rs232_write_dtr %02x\n", BIT(data, 4));
+			m_rs232_conn_dtr_handler(BIT(data, 4));
+		}
+	}
+
+	m_kbdmcu->set_input_line(MCS48_INPUT_IRQ, BIT(data, 7) ? ASSERT_LINE : CLEAR_LINE);
+}
+
+READ8_MEMBER(qvt102_device::kbd_r)
+{
+	// 7------- keyboard to host
+	// -6------ dtr2
+
+	uint8_t data = 0;
+
+	data |= !(m_aux->dsr_r() || m_aux->dcd_r()) << 6;
+	data |= !m_kbd_data << 7;
+
+	return data;
+}
+
+READ8_MEMBER(qvt102_device::ctc_r)
+{
+	return m_ctc->read(BIT(m_latch, 5));
+}
+
+WRITE8_MEMBER(qvt102_device::ctc_w)
+{
+	m_ctc->write(BIT(m_latch, 5), data);
+}
+
+WRITE_LINE_MEMBER(qvt102_device::acia_txd_w)
+{
+	if (BIT(m_latch, 6) == 0)
+	{
+		// print
+		m_aux->write_txd(state);
+	}
+	else
+	{
+		// aux enable
+		if (BIT(m_latch, 0) == 0)
+			m_aux->write_txd(state);
+
+		fprintf(stderr, "rs232_write_txd %02x\n", state);
```

Comment:
Once again, use MAME's diagnostic outputs appropriately, and put stuff that shouldn't be surprising behind `VERBOSE` using `logmacro.h` stuff.